lib/chef-dk/policyfile/git_lock_fetcher.rb

```diff
@@ -0,0 +1,259 @@
+#
+# Copyright:: Copyright (c) 2018 Chef Software Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef-dk/policyfile_lock"
+require "chef-dk/exceptions"
+require "chef-dk/helpers"
+require "mixlib/shellout"
+require "tmpdir"
+
+module ChefDK
+  module Policyfile
+
+    # A Policyfile lock fetcher that can read a lock file from a git repository.
+    #
+    # @author Ryan Hass
+    # @author Daniel DeLeo
+    #
+    # @since 3.1.0
+    #
+    class GitLockFetcher
+      attr_accessor :name
+      attr_accessor :source_options
+      attr_accessor :storage_config
+
+      attr_reader :uri
+      attr_reader :branch
+      attr_reader :tag
+      attr_reader :ref
+      attr_reader :revision
+      attr_reader :path
+
+      # Initialize a GitLockFetcher
+      #
+      # @param name [String] The name of the policyfile
+      # @param source_options [Hash] A hash with a :path key pointing at the location
+      #                              of the lock
+      def initialize(name, source_options, storage_config)
+        @name           = name
+        @storage_config = storage_config
+        @source_options = source_options
+
+        @uri      = source_options[:git]
+        @branch   = source_options[:branch]
+        @tag      = source_options[:tag]
+        @ref      = source_options[:ref]
+        @revision = source_options[:revision]
+        @path     = source_options[:path] || source_options[:rel]
+
+        # The revision to parse
+        @rev_parse = source_options[:ref] || source_options[:branch] || source_options[:tag] || "master"
+      end
+
+      # @return [True] if there were no errors with the provided source_options
+      # @return [False] if there were errors with the provided source_options
+      def valid?
+        errors.empty?
+      end
+
+      # Check the options provided when craeting this class for errors
+      #
+      # @return [Array<String>] A list of errors found
+      def errors
```

Comment:
do you need to check for conflicting information such as someone provided multiple of ref, revision, branch, and tag?